toolchain/check/testdata/operators/overloaded/string_indexing.carbon

```diff
@@ -0,0 +1,299 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/full.carbon
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/operators/overloaded/string_indexing.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/operators/overloaded/string_indexing.carbon
+
+// --- fail_literal_negative_index.carbon
+library "[[@TEST_NAME]]";
+
+fn TestLiteralNegativeIndex() {
+  //@dump-sem-ir-begin
+  // CHECK:STDERR: fail_literal_negative_index.carbon:[[@LINE+4]]:17: error: index `-1` is negative. [StringIndexNegative]
+  // CHECK:STDERR:   let c: char = "Test"[-1];
+  // CHECK:STDERR:                 ^~~~~~~~~~
+  // CHECK:STDERR:
+  let c: char = "Test"[-1];
+  //@dump-sem-ir-end
+}
+
+// --- fail_literal_out_of_bounds.carbon
+library "[[@TEST_NAME]]";
+
+fn TestLiteralOutOfBounds() {
+  //@dump-sem-ir-begin
+  // CHECK:STDERR: fail_literal_out_of_bounds.carbon:[[@LINE+4]]:17: error: string index `4` is out of bounds; string has length 4. [StringAtIndexOutOfBounds]
+  // CHECK:STDERR:   let c: char = "Test"[4];
+  // CHECK:STDERR:                 ^~~~~~~~~
+  // CHECK:STDERR:
+  let c: char = "Test"[4];
+  //@dump-sem-ir-end
+}
+
+// --- fail_wrong_type.carbon
+library "[[@TEST_NAME]]";
+
+fn TestWrongType() {
+  let x: i32 = 42;
+  //@dump-sem-ir-begin
+  // CHECK:STDERR: fail_wrong_type.carbon:[[@LINE+4]]:17: error: cannot access member of interface `Core.IndexWith(Core.IntLiteral)` in type `i32` that does not implement that interface [MissingImplInMemberAccess]
+  // CHECK:STDERR:   let c: char = x[0];
+  // CHECK:STDERR:                 ^~~~
+  // CHECK:STDERR:
+  let c: char = x[0];
+  //@dump-sem-ir-end
+}
+
```

Comment:
This is not testing StringAt, it's testing that there is no `impl i32 as IndexWith`, which doesn't fit here. I think I would just drop this test. We probably don't want to test for the lack of every interface in Core for every type in Core that doesn't implement it.

What about instead testing the wrong type in the argument, like
```
class C {}
"Test"[C];
```